src/pudl/output/ferc1.py

```diff
@@ -2774,3 +2779,239 @@ def nodes_to_df(calc_forest: nx.DiGraph, nodes: list[NodeId]) -> pd.DataFrame:
     except AttributeError:
         tags = pd.DataFrame()
     return pd.concat([index, tags], axis="columns")
+
+
+def _propagate_tags_leafward(
+    annotated_forest: nx.DiGraph, leafward_inherited_tags: list[str]
+) -> nx.DiGraph:
+    existing_tags = nx.get_node_attributes(annotated_forest, "tags")
+    for node, parent_tags in existing_tags.items():
+        descendants = nx.descendants(annotated_forest, node)
+        descendant_tags = {
+            desc: {
+                "tags": {
+                    tag_name: parent_tags[tag_name]
+                    for tag_name in leafward_inherited_tags
+                    if tag_name in parent_tags
+                }
+                | existing_tags.get(desc, {})
+            }
+            for desc in descendants
+        }
+        nx.set_node_attributes(annotated_forest, descendant_tags)
+    return annotated_forest
+
+
+def _propagate_tag_rootward(
```

Comment:
this one is named with `_tag_` instead of `_tags_` rn because it only takes one. if we wanted to keep it consistent i would call this one inside of a `_tags_` function that loops over the tags. but rn this is all only really for the `in_rate_base` tag so that felt premature/necessary.